dali/imgcodec/decoders/nvjpeg_lossless/nvjpeg_lossless.cc

```diff
@@ -0,0 +1,174 @@
+    return is_supported == 0;
+  } catch (...) {
+    return false;
+  }
+}
+
+FutureDecodeResults NvJpegLosslessDecoderInstance::ScheduleDecode(DecodeContext ctx,
+                                                                  span<SampleView<GPUBackend>> out,
+                                                                  cspan<ImageSource *> in,
+                                                                  DecodeParams opts,
+                                                                  cspan<ROI> rois) {
+  int nsamples = in.size();
+  assert(out.size() == nsamples);
+  assert(rois.empty() || rois.size() == nsamples);
+  assert(ctx.tp != nullptr);
+
+  DecodeResultsPromise promise(nsamples);
+  auto set_promise = [&](DecodeResult result) {
+    for (int i = 0; i < nsamples; i++)
+      promise.set(i, result);
+  };
+
+  kernels::DynamicScratchpad s({}, ctx.stream);
+  try {
+    if (opts.format != DALI_ANY_DATA && opts.format != DALI_GRAY)
+      throw std::invalid_argument("Only ANY_DATA and GRAY are supported.");
+
+    sample_meta_.clear();
+    sample_meta_.resize(nsamples);
+    encoded_.clear();
+    encoded_.resize(nsamples);
+    encoded_len_.clear();
+    encoded_len_.resize(nsamples);
+    decoded_.clear();
+    decoded_.resize(nsamples);
+
+    for (int i = 0; i < nsamples; i++) {
+      auto *sample = in[i];
+      auto &out_sample = out[i];
+      assert(sample->Kind() == InputKind::HostMemory);
+      auto *data_ptr = sample->RawData<unsigned char>();
+      auto data_size = sample->Size();
+      encoded_[i] = data_ptr;
+      encoded_len_[i] = data_size;
+      sample_meta_[i].needs_processing = opts.dtype != DALI_UINT16;
+      if (!rois.empty() && rois[i].use_roi()) {
+        sample_meta_[i].needs_processing = true;
+      }
+      if (opts.use_orientation) {
+        auto &ori = sample_meta_[i].orientation = JpegParser().GetInfo(in[i]).orientation;
+        if (ori.rotate || ori.flip_x || ori.flip_y)
+          sample_meta_[i].needs_processing = true;
+      }
+
+      CUDA_CALL(nvjpegJpegStreamParseHeader(nvjpeg_handle_, sample->RawData<unsigned char>(),
+                                            sample->Size(), jpeg_stream_));
+      unsigned int precision;
+      CUDA_CALL(nvjpegJpegStreamGetSamplePrecision(jpeg_stream_, &precision));
+      if (NeedDynamicRangeScaling(precision, DALI_UINT16)) {
+        sample_meta_[i].dyn_range_multiplier = DynamicRangeMultiplier(precision, DALI_UINT16);
+        sample_meta_[i].needs_processing = true;
+      }
+
+      auto &o = decoded_[i];
+      auto sh = out_sample.shape();
+      o.pitch[0] = sh[1] * sh[2] * sizeof(uint16_t);
+      if (sample_meta_[i].needs_processing) {
+        int64_t nbytes = volume(sh) * sizeof(uint16_t);
+        o.channel[0] = s.Allocate<mm::memory_kind::device, uint8_t>(nbytes);
+      } else {
+        o.channel[0] = static_cast<uint8_t *>(out_sample.raw_mutable_data());
```

Comment:
I assume we are guaranteed to have only one component? I haven't seen the doc for this NVJPEG_OUTPUT_UNCHANGEDI_U16 yet.

Reply:
No, it's currently up to 2 components, but in the future it'll be any number of components.
However, channel[i] is only used when decoding to separate planes. When decoding interleaved, only the first element is used (decoded as a single plane with interleaved channels).